tests/test_rfn_query.py

```diff
@@ -0,0 +1,138 @@
+import datetime as dt
+import pandas as pd
+from pandas.compat import lrange
+from pandas import DataFrame
+
+from bgbb.sql.sql_utils import run_rec_freq_spk, to_s3_fmt
+import pyspark
+from pytest import fixture
+
+
+spark = pyspark.sql.SparkSession.builder.appName("test").getOrCreate()
+
+
+def int2date(i):
+    return model_start_date + dt.timedelta(days=i - 1)
+
+
+model_start_date = dt.date(2018, 6, 1)
+MAX_OPPORTUNITIES = 9
+DURATION = 10
+rfn_cols = ["Recency", "Frequency", "N"]
+
+
+def generate_clients_daily():
+    """
+    Testing model window of 10 day.
+    Day 0: day before model start date
+    Day 1: first day of model range
+    Day 10: last day of model range
+    Day 11: first day *after* model range ('holdout start date')
+    """
+    #
+    client_dates_i = [
+        ("Beginning_end", [1, 3, 10]),
+        ("Beginning_end_extra", [0, 1, 3, 10, 11]),
+        (
+            "Before_beginning",
+            [0],
+        ),  # Not included, since this is before model range
+        ("Halfway_start", [5, 6, 11]),
+        ("Every_day", lrange(11)),
+        ("One_time_returner", [3, 4]),
+        ("No_returns", [3]),
+    ]
+
+    client_dates = [
+        (client_name, [int2date(i) for i in ints])
+        for client_name, ints in client_dates_i
+    ]
+
+    clients_daily_df = DataFrame(
+        [
+            (client_name, to_s3_fmt(date))
+            for client_name, dates in client_dates
+            for date in dates
+        ],
+        columns=["client_id", "submission_date_s3"],
+    )
+
+    clients_daily_df["app_name"] = "Firefox"
+    clients_daily_df["channel"] = "release"
+    clients_daily_df["sample_id"] = "1"
+
+    return clients_daily_df
+
+
+def generate_rfn_df():
```

Comment:
If you end up using decorated functions and objects, fixtures can depend on other fixtures.

```suggestion
def generate_rfn_df(spark, generate_clients_daily):
```

Reply:
that...changes everything